scipy/_lib/deprecation.py

```diff
@@ -13,7 +13,7 @@
 _NoValue = object()
 
 def _sub_module_deprecation(*, sub_package, module, private_modules, all,
-                            attribute, correct_module=None):
+                            attribute, correct_module=None, dep_version="1.16.0"):
```

Comment:
I'd prefer not to set a default, because this lead exactly to this kind of situation where we end up forgetting it. IMO it should be a required argument of `_sub_module_deprecation`

Reply:
I considered this but this would require adding a version to all existing uses. Apart from this exception they have all been cleaned up so that seemed like a lot of work for a warning that is not used (except for `interpnd`).

Reply:
But this PR is going to change all those warning messages from 1.14 to 1.16? https://github.com/scipy/scipy/issues/22097 says that one deprecation warning is wrong, but you seem to be saying that all ~90+ [occurrences](https://github.com/search?q=repo%3Ascipy%2Fscipy%20_sub_module_deprecation&type=code) of this function are wrong?

(to be fair, it makes no sense in SciPy 1.15 to say that things will be removed in 1.14, so presumably those _are_ all wrong indeed?)

Reply:
No only one occurrence of the function is wrong, to be clear the situation is:
1. `_sub_module_deprecation` was introduced with a warning about things being removed in 1.14
2. In version 1.14 all the things that were meant to be removed were handled by your pr #19904. At this point the path that lead to the warning mentioning 1.14 was no longer ever reached. This is verified by the following test
https://github.com/scipy/scipy/blob/b1d35063ad50b088ad538d9ffb6b89b0cf17811c/scipy/_lib/tests/test_public_api.py#L445
3. #20612 and #21754 introduced new usage of `_sub_module_deprecation` however mistakenly the version in the warning was not updated. Only the latter pr reaches the 1.14 warning path as reported by @jorenham.

This brings us to this pr which provides a minimal fix to display the correct version number for ease of backporting. Once this is merged, we can immediately remove the offending items from `__all__` and add these two submodules to the above test since development for 1.16 is open. We could also change `_sub_module_deprecation` so that the updating the version is not forgotten again in the future, as requested above.

Reply:
OK, sounds good, thanks (I was just wondering how all the call-sites of `_sub_module_deprecation` would still create correct warning messages, but if indeed those warnings are never reached anymore, then it's obviously OK)